src/common/JSHandle.ts

```diff
@@ -890,15 +887,17 @@ export class ElementHandle<
     ) => ReturnType | Promise<ReturnType>,
     ...args: SerializableOrJSHandle[]
   ): Promise<WrapElementHandle<ReturnType>> {
-    const defaultHandler = (element: Element, selector: string) =>
-      Array.from(element.querySelectorAll(selector));
     const { updatedSelector, queryHandler } = getQueryHandlerAndSelector(
-      selector,
-      defaultHandler
+      selector
     );
-
+    const queryHandlerToArrayStr = `(element, selector) =>
+      Array.from((${queryHandler.queryAll})(element, selector))`;
+    /* The `eval` ensures that the query handler is inlined
+     * before sending the program text to the browser.
+     */
+    const queryHandlerToArray = eval(queryHandlerToArrayStr);
```

Comment:
I'm wondering if we could do something like:

```js
const queryHandlerToArray = Function(
  'element',
  'selector',
  `Array.from((${queryHandler.queryAll})(element, selector))`);
```

Or, even more simply, we could pass the string (`queryHandlerToArrayStr`) directly to `evaluateHandle`. WDYT?

Reply:
+1

Reply:
I actually tried both. `new Function` gives a type error since an `EvaluateHandleFn` is expected, and casting seems to fail since things start getting undefined (I haven't debugged that further though).
Passing the string directly to `evaluateHandle` seems like the correct way forward to me, but the pipeline for evaluating something passed as string is apparently quite different? As far as I can tell, `evaluateHandle` ignores the `args` if `pageFunction` is a string.

Reply:
@johanbay the typescript definitions are a work in progress - so please shout if they cause you problems. There could well be a problem with the types, not with your code. Happy to help with any of this :)

Reply:
@jackfranklin thanks! I was wrong on `Function('element', 'selector', <handlerBody>)` not working here, it actually does. Only problem is that the type `EvaluateHandleFn` clashes with the type `Function`. Do you see any reason not to change `(...args: unknown[]) => unknown` to `Function` in `EvaluateHandleFn` ([link to file](https://github.com/puppeteer/puppeteer/blob/fb806109a09e82b51734615a958dbd1518e0c5e4/src/common/EvalTypes.ts#L40))